homeassistant/components/mqtt/client.py

```diff
@@ -809,6 +811,13 @@ def _mqtt_handle_message(self, msg: mqtt.MQTTMessage) -> None:
         subscriptions = self._matching_subscriptions(msg.topic)
 
         for subscription in subscriptions:
+            init_status = self._retained_init.setdefault(subscription, set())
+            # skip already initialized subscriptions
+            if msg.retain and msg.topic in init_status:
+                continue
+            # remember the subscription had an initial payload
+            self._retained_init[subscription].add(msg.topic)
```

Comment:
This should only happen if msg.retain = true.  The current code will add an entry to the subscription list even if msg.retain = false.

So it should be more like:
```
if msg.retain:
   if msg.topic in init_status:
      continue
   self._retained_init[subscription].add(msg.topic)
```

Also, couldn't the last line just be `init_status.add(msg.topic)` ?

Reply:
The test case for this is:

```
Sub A subscribed.
retain=true msg received.
Sub B subscribed (could be subscription entry created, but not actually subscribed due to debounce)
retain=false msg received
retain=true msg received
```

In the above scenario, the retain=true meant for Sub B would not be processed with the current code.

Reply:
The real case is that when subscribing to an already retained topic, the first message will be a retained one. It is replayed for the new subscription. But...
If your subscribe first and there was nothing to replay and then a retained payload is published, the broker will NOT add the retain flag to the existing subscribers, so you cannot tell if a payload was retained or not unless you are resubscribing.
Hence I believe the replayed and retained payload is redundant for the existing subscriptions. If another device with a different IP would subscribe to the same topic, I would never get the retained payloads replayed if an other device is subscribing. That is weird, but it is what is happening in now, and what I try to solve here.

So for your case: If Sub A gets a retained payload, there must have been a retained payload already at the broker.
If there is not a retained payload a at the broker while subscribing, the next payload received might be retained when publishing but will NOT have a retained flag for existing subscribers as it it just an update for them, and as they were already listening they should already have that update. It is only for the new subscriptions we want the retained payload once for each topic. By storing the topic we can also deal with wildcard subscriptions.

Reply:
TL:DR - Subscribe classes can receive messages on their callbacks before we actually submitted the subscribe to MQTT for that subscribe class, if we have previously subscribed to the same topic.

I'm actually more thinking of the debounce case. 

If we subscribe to Sub A.  It will get it's RETAIN message.  Then start getting retain=0 messages.

Then later (say 5 minutes later) we request to subscribe to Sub B, where we create the Subscribe class, but don't actually submit it to MQTT until the debounce period expires.

In the intervening time, some RETAIN=0 messages can be received (ostensibly for Sub A), but because of the implementation of how received messages are handled, we just cycle through all Subscribe classes for that topic.  We don't first check to see if we've actually sent this to MQTT yet, or gotten the relevant SubAck.

THAT is the case where a Subscribe class can get a RETAIN=0 callback before a RETAIN=1 message.  It's not because of the specification, but because of how it is implemented.  If we were not creating the Subscribe class until we got the SUBACK from MQTT, you would likely be right (though there is still a small possibility that a RETAIN=0 message could be in flight and sent before a RETAIN=1 message is generated/sent - but that's a very small chance).

But we're not waiting for the SUBACK to create the Subscribe class, AND we have now added a further delay between creating the Subscribe class and actually submitting the MQTT SUBSCRIBE (due to debounce).  Which means, due to message processing, a Subscribe class can receive messages (which will all be RETAIN=0) before it's subscribe has even been sent to MQTT.  But it will still get it's RETAIN=1 message when that MQTT SUBSCRIBE is sent and processed by the MQTT server.

Reply:
If we want to be fully compliant to the spec, we could track which MQTT subscribe response is associated with which Subscribe class, and not allow messages until we have gotten the SUBACK for that subscribe.  Though we only wait 10s on the MID before we stop looking for it, and on a busy server, it's not guaranteed you will get a response that quick (if subscribing to a large number of topics, either individually or in a batch - something I've seen personally).

https://github.com/jbouwh/core/blob/mqtt-block-replay-retained/homeassistant/components/mqtt/client.py#L686
https://github.com/jbouwh/core/blob/mqtt-block-replay-retained/homeassistant/components/mqtt/client.py#L901

BUT - unless we are going to wait for the SUBACK before allowing a Subscribe class to process ANY messages, it will be able to receive RETAIN=0 messages before it's RETAIN=1 if we have a prior subscription.

And we HAVE to process the RETAIN=1 message even if we got prior RETAIN=0 messages (once per Subscribe class), as the RETAIN=0 messages might just be deltas, but the RETAIN=1 will be a full 'image'.

And coming back to the current code, it will make a subscription pretend it got a RETAIN=1 message, even if it processed a RETAIN=0 message ('mistakenly' due to the implementation).  Which it shouldn't.

Reply:
Sorry for even more text.

But another possible solution would be this.

Add a `self._has_retained[msg.topic]` so that:

```
if msg.retain and msg.topic not in self._has_retained:
   self._has_retained.add(msg.topic)
```

Then we can add a check of:
```
if not msg.retain and msg.topic in self._has_retained and msg.topic not in self._retained_init[subscription]:
    continue
```
(ie. this is a topic we KNOW uses RETAIN=1 messages, but this subscribe instance has not yet received a RETAIN=1 message, so we should not process any RETAIN=0 messages until we get one).

This would prevent the case of a Subscription class processing any RETAIN=0 messages (due to already being subscribed on the topic with another subscription) before it receives it's RETAIN=1 message (which might be delayed due to debounce).

You would then obviously need to `del self._has_retained[msg.topic]` once we have no more subscriptions for that topic (ie. the same condition we use for actually sending the UNSUBSCRIBE to MQTT).

Reply:
I think the solution I proposed is okay. True, it is possible that we receive updates before we actual resubscribe, and that is not an issue IMO, because we then will ignore any retained messages coming later when the actual subscribe comes and possible retained updates are offered. BTW.

Reply:
But Sub B (the second subscription) should NOT ignore the RETAIN=1 message if it has not gotten one before.

The retain=0 messages should be considered as deltas 'set subfield foo = 5' or even 'set subfield foo += 5'
The retain=1 message would be the full 'frame', (ie. it has all values for all fields).

It may or may not be that way in practice, but that's how it should be considered.  So we can't drop the RETAIN=1 message for Sub B just because it already received a RETAIN=0 message (that it was technically not supposed to get, but oh well).

That's why I said the `self._retained_init[subscription].add(msg.topic)` should only happen if msg.retain = True.

Reply:
Just to wit.  As a publisher, I could publish:
```
1. RETAIN=1 - { field1: 500, field2: "foo", field3: 3.14 }
2. RETAIN=0 - { field2: "bar" }
3. RETAIN=0 - { error: "Some Error Message" }
```

On the client side (first subscriber), I would receive:
```
1. RETAIN=1 - { field1: 500, field2: "foo", field3: 3.14 }
2. RETAIN=0 - { field2: "bar" }
3. RETAIN=0 - { error: "Some Error Message" }
```

If the second subscription class was created after message 1 above, but the subscribe didn't happen until later due to debounce, it would receive:
```
1. RETAIN=0 - { field2: "bar" }
2. RETAIN=0 - { error: "Some Error Message" }
3. RETAIN=1 - { field1: 500, field2: "foo", field3: 3.14 }
```

This means that, if the second subscription doesn't still process the RETAIN=1 message (as with current code), but instead drops it because it already processed a RETAIN=0 message (which is what current code does), it will never have the full picture, until eventually the publisher decided to publish a new full 'image' message (which it would send as RETAIN=1, but our subscribers would receive as RETAIN=0).

Reply:
This is not correct, please try your usecases in real live as I did, and you will see this works differently.

Reply:
Are you saying that I, as an MQTT publisher, am required to send the entire state of the device I'm managing with every message, whether it's RETAIN=1 or RETAIN=0?   That I can't send a partial update?

So if I have 100 fields in the JSON I am sending, I MUST send the ENTIRE JSON with every message?  Or can I, say, decide to send a smaller JSON with only 1 field, or 2 fields, as they change, assuming that the client knows the state of the other 98 or 99 from the RETAIN=1 message they got.

If I've got a value updating, say, 60 times per second or more, I don't want to force you to get 100 values 60 times per second when I can send 1 value 60 times per second.

As far as I know, this is allowed with MQTT.  And is a much more efficient way to USE MQTT.

Reply:
> Are you saying that I, as an MQTT publisher, am required to send the entire state of the device I'm managing with every message, whether it's RETAIN=1 or RETAIN=0? That I can't send a partial update?

You can use multiple topics, but always mus sent the complete payload for that topic. Again, the top of this chain was if we only should initialize if the retain flag is set. The answer is `no` as we should take the first update that is offered (retained or not) an consider this as out best last state, after that we should filter out any retained payloads, as they are trigger by other resubscribes. At reconnect we start over again.